refactor table ddl events scoped for shards

thanodnl opened the pull request:
Refactor internals on how Citus creates the SQL commands it sends to recreate shards.

Before Citus collected solely ddl commands as `char *`'s to recreate a table. If they were used to create a shard they were wrapped with `worker_apply_shard_ddl_command` and send to the workers. On the workers the UDF wrapping the ddl command would rewrite the parsetree to replace tables names with their shard name equivalent.

This worked well, but poses an issue when adding columnar. Due to limitations in Postgres on creating custom options on table access methods we need to fall back on a UDF to set columnar specific options. Now, to recreate the table, we can not longer rely on having solely DDL statements to recreate a table.

A prototype was made to run this UDF wrapped in `worker_apply_shard_ddl_command`. This became pretty messy, hard to understand and subsequently hard to maintain.

This PR proposes a refactor of the internal representation of table ddl commands into a `TableDDLCommand` structure. The current implementation only supports a `char *` as its contents. Based on the use of the DDL statement (eg. creating the table -mx- or creating a shard) one of two different functions can be called to get the statement to send to the worker:
 - `GetTableDDLCommand(TableDDLCommand *command)`: This function returns that ddl command to create the table. In this implementation it will just return the `char *`. This has the same functionality as getting the old list and not wrapping it.
 - `GetShardedTableDDLCommand(TableDDLCommand *command, uint64 shardId, char *schemaName)`: This function returns the ddl command wrapped in `worker_apply_shard_ddl_command` with the `shardId` as an argument. Due to backwards compatibility it also accepts a. `schemaName`. The exact purpose is not directly clear. Ideally new implementations would work with fully qualified statements and ignore the `schemaName`.

A future implementation could accept 2.function pointers and a `void *` for context to let the two pointers work on. This gives greater flexibility in controlling what commands get send in which situations. Also, in a future, we could implement the intermediate step of creating the `parsetree` datastructure of statements based on the contents in the catalog with a corresponding deparser. For sharded queries a mutator could be ran over the parsetree to rewrite the tablenames to the names with the shard identifier. This will completely omit the requirement for `worker_apply_shard_ddl_command`.

TODO:
 - [ ] Enterprise: wrap policy commands in `makeTableDDLCommandString` (awaiting approval of this PR)

thanodnl commented on src/backend/distributed/operations/node_protocol.c:
```diff
@@ -841,3 +851,101 @@ DistributedTableReplicationIsEnabled()
+
+
+char *
+GetTableDDLCommand(TableDDLCommand *command)
```
missing comment

thanodnl commented on src/backend/distributed/operations/node_protocol.c:
```diff
@@ -841,3 +851,101 @@ DistributedTableReplicationIsEnabled()
+
+
+/*
+ * GetShardedTableDDLCommandString is the internal function for TableDDLCommand objects
```
wrong comment

marcocitus commented on src/backend/distributed/operations/node_protocol.c:
```diff
@@ -841,3 +851,101 @@ DistributedTableReplicationIsEnabled()
+	else
+	{
+		appendStringInfo(&buf, WORKER_APPLY_SHARD_DDL_COMMAND_WITHOUT_SCHEMA, shardId,
+						 escapedDDLCommand);
```
I'm not sure why we have this. It seems redundant.

marcocitus commented on src/include/distributed/coordinator_protocol.h:
```diff
@@ -83,6 +83,31 @@ typedef enum
+} TableDDLCommandType;
+
+
+typedef struct TableDDLCommand
```
These structs could use some comments describing what they are for.

I suppose it may be worth noting that this is not strictly for "DDL". (I think the name is good enough, but comment would be helpful)

marcocitus reviewed (approved):
After addressing comments, this seems fine.

thanodnl commented:
have addressed feedback, will merge tomorrow as I want to update enterprise directly with these changes. Enterprise will start failing tests after this merge in the policy area. The policy commands need to be wrapped with `makeTableDDLCommandString(...)`.
